### ISSUE TITLE
Global undo/redo

### ISSUE THREAD

**Issue by tntscreed** [olivierkes/manuskript#1146]
So I've seen this function mentioned in PR https://github.com/olivierkes/manuskript/pull/1113. I don't really have a lot of experience with implementing this in relatively complex software like Manuskript, but I think the only way to do it is with an Action class which is inherited by all possible actions' classes. This means all user-interaction with Manuksript that results in data being modified have to be grouped into their own action classes. These classes would need to have the methods to actually make those changes, and have an opposite action class (with the same, or even more parameters as the original class) that could revert the changes made. Every user action would have to be stored in a list in mainWindow.py.

This would probably not be too easy to implement, and a lot of actions have to be considered and made. Is there a better way of going about this?

**Comment by TheJackiMonster** [comment]
> This would probably not be too easy to implement, and a lot of actions have to be considered and made. Is there a better way of going about this?

No, that's pretty much correct. It's easier to implement by storing multiple states to go back and forth between them. But the downside is huge memory cost. So I don't think that's a valid solution when we have dynamic sizes of text files.

**Comment by tntscreed** [comment]
@TheJackiMonster 
Is it realistic to try to implement this in the QT branch, or even the GTK branch? It seems like a job for a relatively large group of people. Or maybe the bigger problem is the lack of good organisation. It would really have to be planned before jumping into it, and a creative decision has to be made on each action's implementation.

**Comment by TheJackiMonster** [comment]
I mean technically you can try to implement it in an abstract way which might be suitable for both branches. For example if you change the value of a property storing a primitive value or a string, you can just store the old and the new value, some information linking to the object/data and the property name.

I think most difficult actions are deletions, renaming and rearranging structures. You definitely have to plan these before implementing. But maybe such actions can be implemented in an abstract way working for chracters/world/outline in a very similar way.

**Comment by TheJackiMonster** [comment]
The goal should definitely be to keep the code simple and maintainable.

**Comment by TheShadowOfHassen** [comment]
So I was doing some thinking about this. With text libreOffice tracks every word. (I.E a space in between) If we could somehow name each edited data and location it'd be easy. Y
ou track each change in a class storing the UI place and the undo and redo data. You put the classes to a list then set a value for maximun undo/ redo (Fifty?) And you'd have it.

**Comment by obw** [comment]
> So I was doing some thinking about this. With text libreOffice tracks every word. (I.E a space in between) If we could somehow name each edited data and location it'd be easy. Y ou track each change in a class storing the UI place and the undo and redo data. You put the classes to a list then set a value for maximun undo/ redo (Fifty?) And you'd have it.

I think, this is something, we should let the User decide. Most machines nowadays have lots of RAM!

So I think something like 200 or even 1000 is more like it. A normal Manuskript Session is about 8 to 12 hours, (I write about one to two Chapters with about 10thousend up too 20thousend Words!).

Also, my editorial Sessions are shorter, but then I work on four to eight Chapters!

A nice Feature would be, when this Class writes its content on save, so when Manuskript is restarted, the old Undo-History is back again. Manuskript crashes very seldom, but it happens. Also, when I write on the run, I have ignored my low Battery more than once;)

I am a technical Guy, so I would like a Button/Menu Entry to kill the undo memory, rather than having a Max-Element Counter. I also see, why it's better for the most users of manuskript!

For the GUI I think, we should place it where Metadata and so on is placed!

Something like this:

**book > chapter 1 > scene 1**

1. _Change 1 (Timestamp)_
  -  Old
  - New
2. _Change 2 (Timestamp)_
  -  Old
  - New
3. _Change 3 (Timestamp)_
  -  Old
  - New


**book > chapter 1 > scene 2**
...

**Comment by obw** [comment]
> So I was doing some thinking about this. With text libreOffice tracks every word. (I.E a space in between) If we could somehow name each edited data and location it'd be easy. Y ou track each change in a class storing the UI place and the undo and redo data. You put the classes to a list then set a value for maximun undo/ redo (Fifty?) And you'd have it.

I think, this is something, we should let the User decide. Most machines nowadays have lots of RAM!

So I think something like 200 or even 1000 is more like it. A normal Manuskript Session is about 8 to 12 hours, (I write about one to two Chapters with about 10thousend up too 20thousend Words!).

Also, my editorial Sessions are shorter, but then I work on four to eight Chapters!

A nice Feature would be, when this Class writes its content on save, so when Manuskript is restarted, the old Undo-History is back again. Manuskript crashes very seldom, but it happens. Also, when I write on the run, I have ignored my low Battery more than once;)

I am a technical Guy, so I would like a Button/Menu Entry to kill the undo memory, rather than having a Max-Element Counter. I also see, why it's better for the most users of manuskript!

For the GUI I think, we should place it where Metadata and so on is placed!

Something like this:

**book > chapter 1 > scene 1**

1. _Change 1 (Timestamp)_
  -  Old
  - New
2. _Change 2 (Timestamp)_
  -  Old
  - New
3. _Change 3 (Timestamp)_
  -  Old
  - New


**book > chapter 1 > scene 2**
...

**Comment by TheJackiMonster** [comment]
@obw I think this is where I would like to distinguish between revisions and undo/redo. Undo/Redo should only be implemented to reverse the last few changes which were not saved yet as revision state. Users would still have the option to switch between revision states from the current or previous sessions.

Also keep in mind that we can't implement a system which relies on words if we want to support many international languages. Multiple Asian languages rely on characters instead of words which means we need to track each character being changed to provide a good experience with undo/redo actions.

Theoretically memory is not the biggest issue here but the complexity of the implementation. We don't just track changes inside text files but also deletions, additions or metadata changes of whole files. Those can certainly not be optimized as much as in-text changes.

For GUI I would think we don't need a whole visual log of the undo/redo stacks. It's fine to only show a hint what the last action was a user can undo or redo. The revisions however should get a visual list with timestamps and optional names.

**Comment by TheShadowOfHassen** [comment]
So what I meant by words is no one would want to sit there and undo every word changed so LibreOffice squashes(?) the changes into each word. I had no idea about Asian languages, however changing something for them wouldn't be hard, I think.

The basic premise would be pretty easy. You'd have some sort of tracker attached to each data model and also something that jumps it to the suspific UI place. It would though be a mess because of the hundreds of UI and data elements you have to track.

**Comment by obw** [comment]
@TheJackiMonster:
> @obw I think this is where I would like to distinguish between revisions and undo/redo. Undo/Redo should only be implemented to reverse the last few changes which were not saved yet as revision state. Users would still have the option to switch between revision states from the current or previous sessions.

I don't see this as replacement of a proper, working revision-system, more as a comfort feature.

Also, whatever we are doing with the revision-system, it could be a start for it, when we use this, to pre collect the Data an every time something falls out of the Undo/Redo Buffer, it could be merged with all Edits of the same Element, above it and remove them until, another Element is found.

So the growth of the Revision is smaller, and it should be faster.
> 
> Also keep in mind that we can't implement a system which relies on words if we want to support many international languages. Multiple Asian languages rely on characters instead of words which means we need to track each character being changed to provide a good experience with undo/redo actions.

I know what I think, we should work with ticks, like Games, every X Seconds with changes, store the changes, when they are at the same Place! Also, every time the User changes the Place, a new empty Entry is generated, when the actual first Entry is not Empty.
> 
> Theoretically memory is not the biggest issue here but the complexity of the implementation. We don't just track changes inside text files but also deletions, additions or metadata changes of whole files. Those can certainly not be optimized as much as in-text changes.

That's true, that is the thing, we needed most, it's also the thing which the hardest to implement!
> ...

@TheShadowOfHassen 
> The basic premise would be pretty easy. You'd have some sort of tracker attached to each data model and also something that jumps it to the suspific UI place. It would though be a mess because of the hundreds of UI and data elements you have to track.

The Question to ask where we attach it, if we attach the Undo to the Content-Elements, we have a unique Undo Buffer for each Element. Which is the way, most Users will expect this behavior.

If we make it the other way around, that what I suggested, will be needed!

It's simply a question of preference, I see my Stories always as one homogenous dataset. So I would prefer one UNDO/REDO Storage, which stores the context. Which is not needed, when we attach, this functionality, to each element!

Regards

**Comment by TheShadowOfHassen** [comment]
I don't mean having it attached to each element, but global. If I hit cntrl-z first it undos my delete of the chapter and if I do it again my character's name gets changed back.
